File: Network/NetObjective-c/NetObjective-c/NetworkManager.m
//
//  NetworkManager.m
//  NetObjective-c
//
//  Created by 휴빌론 on 06/02/2020.
//  Copyright © 2020 hubilon. All rights reserved.
//

#import "NetworkManager.h"

@implementation NetworkManager


- (void)getRequestHTTP:(NSString *)url httpMethod:(NSString *)httpMethod completion:(ResponseBlock)completion
{
    NSURL *URL = [NSURL URLWithString:url];
    NSMutableURLRequest *request =[[NSMutableURLRequest alloc] init];
    
//    NSData *postData = [/*보낼 파라메터*/ dataUsingEncoding:NSUTF8StringEncoding];
//    NSString *postLength = [NSString stringWithFormat:@"%d", (int)[postData length]];
//    [request setValue:postLength forHTTPHeaderField:@"Content-Length"];
//    [request setHTTPBody:postData];
    
    [request setURL:URL];
    [request setHTTPMethod:httpMethod];
    [request setValue:@"Mozilla/4.0 (compatible;)" forHTTPHeaderField:@"User-Agent"];
    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                    completionHandler:^(NSData *data, NSURLResponse *response, NSError *error)
                                          {
                                              NSError *myError = error;
                                              NSMutableDictionary *productionSchedule = [[NSMutableDictionary alloc]initWithDictionary:[NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&myError]];
        
                                            /*네트워크 통신이 완료되면 데이터를 콜백한다.*/
                                            completion(productionSchedule,error);
                                              
                                          }];
        [dataTask resume];

}
    
- (void)postRequestHTTP:(NSString *)url httpMethod:(NSString *)httpMethod  body:(NSDictionary *)body completion:(ResponseBlock)completion
{
    NSError *error;

//    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sharedSession];
    NSURL *URL = [NSURL URLWithString:url];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:60.0];

    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request addValue:@"application/json" forHTTPHeaderField:@"Accept"];

    [request setHTTPMethod:@"POST"];
    // NSDictionary *mapData = [[NSDictionary alloc] initWithObjectsAndKeys: @"TEST IOS", @"name",@"IOS TYPE", @"typemap",nil];
    NSDictionary *mapData = body;
    NSData *postData = [NSJSONSerialization dataWithJSONObject:mapData options:0 error:&error];
    [request setHTTPBody:postData];


    NSURLSessionDataTask *postDataTask = [session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
     NSError *myError = error;
     NSMutableDictionary *productionSchedule = [[NSMutableDictionary alloc]initWithDictionary:[NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&myError]];
        /*네트워크 통신이 완료되면 데이터를 콜백한다.*/
        completion(productionSchedule,error);
    }];

    [postDataTask resume];
}
@end

File: Network/NetSwift/NetSwift/NetworkManager.swift
//
//  NetworkManager.swift
//  NetSwift
//
//  Created by 휴빌론 on 06/02/2020.
//  Copyright © 2020 hubilon. All rights reserved.
//

import UIKit

class NetworkManager: NSObject {
    func getRequestHTTP(url: String, type: String, parameter: String? = nil, completion: @escaping(Bool,Dictionary<String, Any>?, Error?) -> Void) {
        
         var urlLink = URLComponents()
            if let query = parameter {
                urlLink = URLComponents(string: url + query)!
            }else{
                urlLink = URLComponents(string: url)!
            }

            // 1. 요청 객체 생성
            var request = URLRequest(url: urlLink.url!)
            
            // 2. 요청 방식
            request.httpMethod = type
            // 3. 헤더
            request.addValue("application/json", forHTTPHeaderField: "Content-Type")

            // 4. Session
            let session = URLSession.shared
            let task = session.dataTask(with: request) { (data, response, error) in
                guard error == nil else {
                    print(error!.localizedDescription)
                    return
                }
                do {
                    let anyData = try JSONSerialization.jsonObject(with: data!, options: [])
                    guard let dic = anyData as? Dictionary<String,Any> else { return }
                    /* completion 코드 삽입 */
                } catch {
                    /* fail 코드 삽입 */
                }
                
            }
            // 6. 실행
            task.resume()
        }

        func postRequestHTTP(url: String, type: String, body: [String:Any], completion: @escaping(Bool,Dictionary<String, Any>?, Error?) -> Void) {
            guard let url = URL(string: url) else { return }
            
            // 1. 요청 객체 생성
            var request = URLRequest(url: url)
            // 2. 요청 방식
            request.httpMethod = type
            // 3. 헤더
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            // 4.바디에 담아서 보낼 데이터 json 형태로 변환
            let jsonData = try? JSONSerialization.data(withJSONObject: body, options:.prettyPrinted)
            request.httpBody = jsonData
            
            
            // 5. Session
            let session = URLSession.shared
            let task = session.dataTask(with: request) { (data, response, error) in
                guard error == nil else {
                    print(error!.localizedDescription)
                    return
                }
                do {
                    let anyData = try JSONSerialization.jsonObject(with: data!, options: [])
                    guard let dic = anyData as? Dictionary<String, Any> else { return }
                    /* completion 코드 삽입 */
                } catch {
                    /* fail 코드 삽입 */
                }
            }
            // 6. 실행
            task.resume()
        }
}
